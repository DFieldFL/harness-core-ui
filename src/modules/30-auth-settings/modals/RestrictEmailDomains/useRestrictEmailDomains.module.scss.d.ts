/* eslint-disable */
// this is an auto-generated file
declare const styles: {
  readonly dialog: string
  readonly heading: string
}
export default styles
