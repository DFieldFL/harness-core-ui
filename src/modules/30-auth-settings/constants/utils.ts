export enum AuthenticationMechanisms {
  SAML = 'SAML',
  OAUTH = 'OAUTH',
  LDAP = 'LDAP',
  USER_PASSWORD = 'USER_PASSWORD'
}

export const getSamlEndpoint = (accountId: string): string => {
  let url = window.location.href.split('/#')[0]?.replace('/ng', '')

  if (window.apiUrl) {
    if (window.apiUrl.startsWith('/')) {
      url = `${url}${window.apiUrl}`
    } else {
      url = window.apiUrl
    }
  }

  return `${url}/api/users/saml-login?accountId=${accountId}`
}
