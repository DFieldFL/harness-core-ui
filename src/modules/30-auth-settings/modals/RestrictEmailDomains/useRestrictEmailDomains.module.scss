.dialog {
  border-radius: 4px;
  width: 560px;

  .heading {
    font-size: 24px !important;
  }
}
