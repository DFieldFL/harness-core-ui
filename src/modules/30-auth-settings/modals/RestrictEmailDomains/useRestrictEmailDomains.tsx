import React from 'react'
import cx from 'classnames'
import { useModalHook } from '@wings-software/uicore'
import { Dialog, Classes } from '@blueprintjs/core'
import RestrictEmailDomainsForm from './views/RestrictEmailDomainsForm'
import css from './useRestrictEmailDomains.module.scss'

interface Props {
  onSuccess: () => void
  whitelistedDomains: string[]
}
interface ModalReturn {
  openRestrictEmailDomainsModal: () => void
  closeRestrictEmailDomainsModal: () => void
}

export const useRestrictEmailDomains = ({ onSuccess, whitelistedDomains }: Props): ModalReturn => {
  const [showModal, hideModal] = useModalHook(
    () => (
      <Dialog isOpen title="" onClose={hideModal} className={cx(css.dialog, Classes.DIALOG)}>
        <RestrictEmailDomainsForm
          onSubmit={() => {
            onSuccess()
            hideModal()
          }}
          onCancel={() => {
            hideModal()
          }}
          whitelistedDomains={whitelistedDomains}
        />
      </Dialog>
    ),
    [whitelistedDomains]
  )

  return {
    openRestrictEmailDomainsModal: showModal,
    closeRestrictEmailDomainsModal: hideModal
  }
}
