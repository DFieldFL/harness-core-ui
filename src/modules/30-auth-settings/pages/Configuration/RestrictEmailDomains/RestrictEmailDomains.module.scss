.card {
  width: 100%;
}

.input {
  margin: var(--spacing-medium) var(--spacing-xxlarge) var(--spacing-small);
  background-color: white !important ;
  border: 1px solid var(--grey-200);
}
