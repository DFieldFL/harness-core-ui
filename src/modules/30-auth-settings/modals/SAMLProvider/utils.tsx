import type { IconName } from '@wings-software/uicore'
import { AuthenticationMechanisms } from '@auth-settings/constants/utils'
import type { UseStringsReturn } from 'framework/strings'

export interface FormValues {
  displayName: string
  authorizationEnabled: boolean
  groupMembershipAttr: string
}

export enum Providers {
  AZURE = 'AZURE',
  OKTA = 'OKTA',
  ONE_LOGIN = 'ONE_LOGIN',
  OTHER = 'OTHER'
}

export interface SAMLProviderType {
  value: Providers
  label: string
  icon: IconName
}

export const createFormData = (data: FormValues): FormData => {
  const formData = new FormData()
  formData.set('displayName', data.displayName)
  formData.set('authorizationEnabled', JSON.stringify(data.authorizationEnabled))
  formData.set('groupMembershipAttr', data.groupMembershipAttr)
  formData.set('ssoSetupType', AuthenticationMechanisms.SAML)

  const file = (data as any)?.files?.[0]
  file && formData.set('file', file)

  return formData
}

export const getSelectedSAMLProvider = (
  selected: SAMLProviderType | undefined,
  getString: UseStringsReturn['getString']
): string => {
  if (selected && selected?.value !== Providers.OTHER) {
    return selected.label
  }
  return getString('authSettings.SAMLProvider')
}
