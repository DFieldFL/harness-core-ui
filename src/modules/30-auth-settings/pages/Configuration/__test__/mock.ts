import type { RestResponseBoolean, LoginSettings } from 'services/cd-ng'
import { AuthenticationMechanisms } from '@auth-settings/constants/utils'
import { ResourceType } from '@common/interfaces/ResourceInterface'

export const mockResponse: RestResponseBoolean = {
  metaData: {},
  resource: true,
  responseMessages: []
}

export const loginSettings: LoginSettings = {
  uuid: '123',
  accountId: '123',
  lastUpdatedAt: 1617864012452,
  userLockoutPolicy: {
    enableLockoutPolicy: true,
    numberOfFailedAttemptsBeforeLockout: 6,
    lockOutPeriod: 24,
    notifyUser: true
  },
  passwordExpirationPolicy: {
    enabled: true,
    daysBeforePasswordExpires: 30,
    daysBeforeUserNotifiedOfPasswordExpiration: 3
  },
  passwordStrengthPolicy: {
    enabled: true,
    minNumberOfCharacters: 12,
    minNumberOfUppercaseCharacters: 1,
    minNumberOfLowercaseCharacters: 1,
    minNumberOfSpecialCharacters: 1,
    minNumberOfDigits: 1
  }
}

export const authSettings = {
  ngAuthSettings: [
    {
      loginSettings,
      settingsType: AuthenticationMechanisms.USER_PASSWORD
    },
    {
      allowedProviders: ['GITHUB', 'AZURE', 'LINKEDIN', 'BITBUCKET', 'GOOGLE'],
      settingsType: AuthenticationMechanisms.OAUTH
    }
  ],
  whitelistedDomains: [],
  authenticationMechanism: AuthenticationMechanisms.USER_PASSWORD,
  twoFactorEnabled: false
}

export const permissionRequest = {
  resourceScope: {
    accountIdentifier: 'accountId'
  },
  resource: {
    resourceType: ResourceType.AUTHSETTING
  }
}
