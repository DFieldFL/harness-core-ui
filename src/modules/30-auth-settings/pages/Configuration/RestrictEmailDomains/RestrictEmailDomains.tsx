import React from 'react'
import { useParams } from 'react-router-dom'
import { Container, Card, Switch, Text, Color } from '@wings-software/uicore'
import { TagInput } from '@blueprintjs/core'
import { useStrings } from 'framework/strings'
import type { AccountPathProps } from '@common/interfaces/RouteInterfaces'
import { useToaster } from '@common/components'
import { useUpdateWhitelistedDomains } from 'services/cd-ng'
import { useRestrictEmailDomains } from '@auth-settings/modals/RestrictEmailDomains/useRestrictEmailDomains'
import { useConfirmationDialog } from '@common/modals/ConfirmDialog/useConfirmationDialog'
import RbacButton from '@rbac/components/Button/Button'
import { PermissionIdentifier } from '@rbac/interfaces/PermissionIdentifier'
import type { PermissionRequest } from '@auth-settings/pages/Configuration/Configuration'
import css from './RestrictEmailDomains.module.scss'

interface Props {
  whitelistedDomains: string[]
  refetchAuthSettings: () => void
  permissionRequest: PermissionRequest
  canEdit: boolean
}

const RestrictEmailDomains: React.FC<Props> = ({
  whitelistedDomains,
  refetchAuthSettings,
  canEdit,
  permissionRequest
}) => {
  const { getString } = useStrings()
  const { showSuccess, showError } = useToaster()
  const { accountId } = useParams<AccountPathProps>()
  const emailRestrictionsEnabled = !!whitelistedDomains.length

  const { mutate: updateWhitelistedDomains, loading: updatingWhitelistedDomains } = useUpdateWhitelistedDomains({
    queryParams: {
      accountIdentifier: accountId
    }
  })

  const onSuccess = (): void => {
    refetchAuthSettings()
  }

  const { openRestrictEmailDomainsModal } = useRestrictEmailDomains({ onSuccess, whitelistedDomains })

  const disableWhitelistedDomains = async (): Promise<void> => {
    try {
      const disabled = await updateWhitelistedDomains([])

      /* istanbul ignore else */ if (disabled) {
        refetchAuthSettings()
        showSuccess(getString('authSettings.whitelistedDomainsDisabled'), 5000)
      }
    } catch (e) {
      /* istanbul ignore next */ showError(e.data?.message || e.message, 5000)
    }
  }

  const { openDialog: confirmWhitelistedDomainsDisable } = useConfirmationDialog({
    titleText: getString('authSettings.disableWhitelistedDomains'),
    contentText: getString('authSettings.confirmDisableWhitelistedDomains'),
    confirmButtonText: getString('confirm'),
    cancelButtonText: getString('cancel'),
    onCloseDialog: isConfirmed => {
      /* istanbul ignore else */ if (isConfirmed) {
        disableWhitelistedDomains()
      }
    }
  })

  const onChangeWhitelistedDomains = (e: React.FormEvent<HTMLInputElement>): void => {
    const enable = e.currentTarget.checked

    if (emailRestrictionsEnabled && !enable) {
      confirmWhitelistedDomainsDisable()
    } else if (!emailRestrictionsEnabled && enable) {
      openRestrictEmailDomainsModal()
    }
  }

  return (
    <Container margin="xlarge">
      <Card className={css.card}>
        <Switch
          labelElement={
            <Text inline color={Color.BLACK} font={{ weight: 'bold', size: 'normal' }}>
              {getString('authSettings.allowUsersWIthEmails')}
            </Text>
          }
          checked={emailRestrictionsEnabled}
          onChange={onChangeWhitelistedDomains}
          disabled={!canEdit || updatingWhitelistedDomains}
          data-testid="toggle-restrict-email-domains"
        />
        {emailRestrictionsEnabled && (
          <TagInput
            disabled
            values={whitelistedDomains}
            rightElement={
              <RbacButton
                minimal
                intent="primary"
                icon="edit"
                onClick={openRestrictEmailDomainsModal}
                disabled={!canEdit}
                data-testid="update-restrict-email-domains"
                permission={{
                  ...permissionRequest,
                  permission: PermissionIdentifier.EDIT_AUTHSETTING
                }}
              />
            }
            className={css.input}
            tagProps={{ minimal: true }}
          />
        )}
      </Card>
    </Container>
  )
}

export default RestrictEmailDomains
