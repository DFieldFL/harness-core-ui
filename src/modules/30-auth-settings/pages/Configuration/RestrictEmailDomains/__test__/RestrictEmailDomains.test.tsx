import React from 'react'
import { render, act, fireEvent, waitFor, queryByText, queryAllByText } from '@testing-library/react'
import { TestWrapper, findDialogContainer } from '@common/utils/testUtils'
import routes from '@common/RouteDefinitions'
import { accountPathProps } from '@common/utils/routeUtils'
import { mockResponse } from '@auth-settings/pages/Configuration/__test__/mock'
import RestrictEmailDomains from '../RestrictEmailDomains'

jest.mock('services/cd-ng', () => ({
  useUpdateWhitelistedDomains: jest.fn().mockImplementation(() => {
    return { mutate: () => Promise.resolve(mockResponse) }
  })
}))

const refetchAuthSettings = jest.fn()
const setUpdating = jest.fn()
const whitelistedDomains = ['harness.io']

describe('RestrictEmailDomains', () => {
  test('Disable email restrictions', async () => {
    const { container, getByTestId } = render(
      <TestWrapper
        path={routes.toAuthenticationSettings({ ...accountPathProps })}
        pathParams={{ accountId: 'testAcc' }}
      >
        <RestrictEmailDomains
          whitelistedDomains={whitelistedDomains}
          refetchAuthSettings={refetchAuthSettings}
          canEdit
          setUpdating={setUpdating}
        />
      </TestWrapper>
    )

    expect(container).toMatchSnapshot()

    const toggleRestrictEmailDomains = getByTestId('toggle-restrict-email-domains')
    act(() => {
      fireEvent.click(toggleRestrictEmailDomains)
    })

    await waitFor(() => queryByText(document.body, 'authSettings.disableWhitelistedDomains'))
    const confirmForm = findDialogContainer()
    expect(confirmForm).toBeTruthy()

    const confirmButton = queryByText(confirmForm!, 'confirm')
    await act(async () => {
      fireEvent.click(confirmButton!)
    })

    expect(queryAllByText(document.body, 'authSettings.whitelistedDomainsDisabled')).toBeTruthy()
  }),
    test('Update email restrictions', async () => {
      const { container, getByTestId } = render(
        <TestWrapper
          path={routes.toAuthenticationSettings({ ...accountPathProps })}
          pathParams={{ accountId: 'testAcc' }}
        >
          <RestrictEmailDomains
            whitelistedDomains={whitelistedDomains}
            refetchAuthSettings={refetchAuthSettings}
            canEdit
            setUpdating={setUpdating}
          />
        </TestWrapper>
      )

      expect(container).toMatchSnapshot()

      const updateRestrictEmailDomains = getByTestId('update-restrict-email-domains')
      act(() => {
        fireEvent.click(updateRestrictEmailDomains)
      })

      await waitFor(() => queryByText(document.body, 'authSettings.allowLoginFromTheseDomains'))
      const form = findDialogContainer()
      expect(form).toBeTruthy()

      const saveButton = queryByText(form!, 'save')
      await act(async () => {
        fireEvent.click(saveButton!)
      })

      expect(queryAllByText(document.body, 'authSettings.WhitelistedDomainsUpdated')).toBeTruthy()
    })
})
