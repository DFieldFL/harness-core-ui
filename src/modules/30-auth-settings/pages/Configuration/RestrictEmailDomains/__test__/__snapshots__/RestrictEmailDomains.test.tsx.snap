// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RestrictEmailDomains Disable email restrictions 1`] = `
<div>
  <div
    class="StyledProps--font StyledProps--main StyledProps--margin-xlarge"
  >
    <div
      class="bp3-card bp3-elevation-0 Card--card card"
    >
      <label
        class="bp3-control bp3-switch StyledProps--font Switch--switch StyledProps--main"
      >
        <input
          checked=""
          data-testid="toggle-restrict-email-domains"
          type="checkbox"
        />
        <span
          class="bp3-control-indicator"
        />
        <span
          class="StyledProps--font StyledProps--main StyledProps--inline StyledProps--color StyledProps--color-black StyledProps--font-weight-bold StyledProps--font-size-normal"
        >
          authSettings.allowUsersWithEmails
        </span>
      </label>
      <div
        class="bp3-input bp3-tag-input bp3-disabled input"
      >
        <div
          class="bp3-tag-input-values"
        >
          <span
            class="bp3-tag bp3-minimal"
            data-tag-index="0"
          >
            <span
              class="bp3-text-overflow-ellipsis bp3-fill"
            >
              harness.io
            </span>
          </span>
          <input
            class="bp3-input-ghost"
            disabled=""
            value=""
          />
        </div>
        <button
          class="bp3-button Button--button StyledProps--font StyledProps--main Button--with-current-color Button--iconOnly Button--withLeftIcon Button--variation Button--variation-icon"
          data-testid="update-restrict-email-domains"
          type="button"
        >
          <span
            class="bp3-icon StyledProps--main"
            data-icon="Edit"
          >
            <svg
              height="16"
              viewBox="0 0 12 12"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g
                fill="none"
                fill-rule="evenodd"
                stroke-linecap="round"
                stroke-linejoin="round"
              >
                <path
                  d="M8.66 1.823h0L2.244 8.239v1.75h1.75l6.416-6.416a1.75 1.75 0 00-1.75-1.75z"
                  stroke="#555770"
                />
                <path
                  d="M6.583 5.146l2.063-2.063"
                  opacity="0.3"
                  stroke="#606170"
                />
              </g>
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`RestrictEmailDomains Update email restrictions 1`] = `
<div>
  <div
    class="StyledProps--font StyledProps--main StyledProps--margin-xlarge"
  >
    <div
      class="bp3-card bp3-elevation-0 Card--card card"
    >
      <label
        class="bp3-control bp3-switch StyledProps--font Switch--switch StyledProps--main"
      >
        <input
          checked=""
          data-testid="toggle-restrict-email-domains"
          type="checkbox"
        />
        <span
          class="bp3-control-indicator"
        />
        <span
          class="StyledProps--font StyledProps--main StyledProps--inline StyledProps--color StyledProps--color-black StyledProps--font-weight-bold StyledProps--font-size-normal"
        >
          authSettings.allowUsersWithEmails
        </span>
      </label>
      <div
        class="bp3-input bp3-tag-input bp3-disabled input"
      >
        <div
          class="bp3-tag-input-values"
        >
          <span
            class="bp3-tag bp3-minimal"
            data-tag-index="0"
          >
            <span
              class="bp3-text-overflow-ellipsis bp3-fill"
            >
              harness.io
            </span>
          </span>
          <input
            class="bp3-input-ghost"
            disabled=""
            value=""
          />
        </div>
        <button
          class="bp3-button Button--button StyledProps--font StyledProps--main Button--with-current-color Button--iconOnly Button--withLeftIcon Button--variation Button--variation-icon"
          data-testid="update-restrict-email-domains"
          type="button"
        >
          <span
            class="bp3-icon StyledProps--main"
            data-icon="Edit"
          >
            <svg
              height="16"
              viewBox="0 0 12 12"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g
                fill="none"
                fill-rule="evenodd"
                stroke-linecap="round"
                stroke-linejoin="round"
              >
                <path
                  d="M8.66 1.823h0L2.244 8.239v1.75h1.75l6.416-6.416a1.75 1.75 0 00-1.75-1.75z"
                  stroke="#555770"
                />
                <path
                  d="M6.583 5.146l2.063-2.063"
                  opacity="0.3"
                  stroke="#606170"
                />
              </g>
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`RestrictEmailDomains Update email restrictions 2`] = `
<div>
  <div
    class="StyledProps--font StyledProps--main StyledProps--margin-xlarge"
  >
    <div
      class="bp3-card bp3-elevation-0 Card--card card"
    >
      <label
        class="bp3-control bp3-switch StyledProps--font Switch--switch StyledProps--main"
      >
        <input
          checked=""
          data-testid="toggle-restrict-email-domains"
          type="checkbox"
        />
        <span
          class="bp3-control-indicator"
        />
        <span
          class="StyledProps--font StyledProps--main StyledProps--inline StyledProps--color StyledProps--color-black StyledProps--font-weight-bold StyledProps--font-size-normal"
        >
          authSettings.allowUsersWithEmails
        </span>
      </label>
      <div
        class="bp3-input bp3-tag-input bp3-disabled input"
      >
        <div
          class="bp3-tag-input-values"
        >
          <span
            class="bp3-tag bp3-minimal"
            data-tag-index="0"
          >
            <span
              class="bp3-text-overflow-ellipsis bp3-fill"
            >
              harness.io
            </span>
          </span>
          <input
            class="bp3-input-ghost"
            disabled=""
            value=""
          />
        </div>
        <button
          class="bp3-button Button--button StyledProps--font StyledProps--main Button--with-current-color Button--iconOnly Button--withLeftIcon Button--variation Button--variation-icon"
          data-testid="update-restrict-email-domains"
          type="button"
        >
          <span
            class="bp3-icon StyledProps--main"
            data-icon="Edit"
          >
            <svg
              height="16"
              viewBox="0 0 12 12"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <g
                fill="none"
                fill-rule="evenodd"
                stroke-linecap="round"
                stroke-linejoin="round"
              >
                <path
                  d="M8.66 1.823h0L2.244 8.239v1.75h1.75l6.416-6.416a1.75 1.75 0 00-1.75-1.75z"
                  stroke="#555770"
                />
                <path
                  d="M6.583 5.146l2.063-2.063"
                  opacity="0.3"
                  stroke="#606170"
                />
              </g>
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
